src/matmuladd.jl

```diff
@@ -0,0 +1,219 @@
+            mul_val = lhs(m)
+
+            if nameof(mul_val) in nameof.(plus_args)
+                A, B = arguments(rhs(m))
+                Cs = filter(x -> nameof(x) != nameof(mul_val), plus_args)
+                validate_mul_shapes(A, B, Cs...) || return nothing
+                return (; A, B, Cs, mul_candidate = m, plus_candidate = c, mul_idx, plus_idx, pattern="A*B + C")
+            end
+        end
+        filter(!isnothing, mul_pattern)
+    end
+    isempty(pattern) ? nothing : pattern
+end
+
+function transform_to_mul5_assignment(expr, match_data_, state::Code.CSEState)
+    Cset = Set(filter(!is_cse_var, reduce(vcat,getproperty.(match_data_, :Cs))))
+    plus_candidates_idx = getproperty.(match_data_, :plus_idx)
+
+    final_temps = []
+
+    m_ = map(match_data_) do match_data
+
+        A, B = match_data.A, match_data.B
+        C = pop!(Cset)
+        T = vartype(C)
+
+        # Create temporary variable for the result
+        temp_var_sym = gensym("mul5_temp")
+        temp_var = Sym{T}(temp_var_sym; type=symtype(C))
+
+        copy_call = Term{T}(copy, [C]; type=symtype(C))
+        mul_call = Term{T}(LinearAlgebra.mul!,
+            [temp_var, A, B, Const{T}(1), Const{T}(1)];
+            type=symtype(C))
+
+        # Add assignments to CSE state
+        copy_assignment = Assignment(temp_var, copy_call)
+        mul_assignment = Assignment(temp_var, mul_call)  # This overwrites temp_var with mul! result
+        final_assignment = Assignment(temp_var, temp_var)
+        push!(final_temps, temp_var)
+
+        [copy_assignment, mul_assignment, final_assignment]
+    end
+    m = m_ |> Base.Fix1(reduce, vcat)
+
+    transformed_idxs = getproperty.(match_data_, :plus_idx)
+    substitution_map = get_substitution_map(match_data_, m_)
+    rm_idxs = getproperty.(match_data_, :mul_idx)
+    transformations = Dict()
+    map(transformed_idxs, m_) do i, mm
+        bank(transformations, i, mm)
+    end
+
+    new_pairs = []
+    for (i, e) in enumerate(expr.pairs)
+        if i in transformed_idxs
+            push!(new_pairs, transformations[i]...)
+            @show e
+        elseif i in rm_idxs
+            push!(new_pairs, nothing)
+        else
+            push!(new_pairs, e)
+        end
+    end
+    new_pairs = filter(!isnothing, new_pairs)
+
+    push!(state.sorted_exprs, m...)
+    temp_var = last(m).lhs
+    new_let = Code.Let(new_pairs, expr.body, expr.let_block)
+    apply_substitution_map(new_let, substitution_map)
+end
+
+function get_substitution_map(match_data, transformations)
+    dic = Dict()
+    @assert length(match_data) == length(transformations)
+
+    plus_idxs = getproperty.(match_data, :plus_idx)
+
+    map(match_data, transformations) do m, t
+        bank(dic, m.plus_candidate.lhs, t[end].lhs)
+    end
+    dic
+end
+
+function bank(dic, key, value)
+    if haskey(dic, key)
+        dic[key] = vcat(dic[key], value)
+    else
+        dic[key] = value
```

Comment:
```suggestion
        push!(dic[key], value)
    else
        dic[key] = copy(value)
```

Allocates once instead of every time a key is pushed.